src/process/agent_exec.rs

```diff
@@ -25,10 +33,145 @@ fn command_with_context(asset_agent_id: &str, command: &str) -> String {
 }
 
-#[cfg(target_os = "windows")]
-pub fn command_execution(asset_agent_id: &str, raw_command: &str) -> Result<(), Error> {
-    use std::os::windows::process::CommandExt;
+// Helper function to get a readable error message from the Windows API
+fn get_last_error_message() -> String {
+    let error_code = unsafe { GetLastError() };
+    if error_code == 0 {
+        return "No error".to_string();
+    }
+
+    let mut buffer: Vec<u16> = Vec::with_capacity(512);
+    let len = unsafe {
+        FormatMessageW(
+            FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_IGNORE_INSERTS,
+            null_mut(),
+            error_code,
+            0,
+            buffer.as_mut_ptr(),
+            buffer.capacity() as u32,
+            null_mut(),
+        )
+    };
+
+    if len == 0 {
+        return "Failed to format error message".to_string();
+    }
+
+    unsafe { buffer.set_len(len as usize) }; // Set the length of the buffer
+
+    // Convert wide string to UTF-8
+    OsString::from_wide(&buffer).to_string_lossy().to_string()
+}
+
+// Helper function to convert a Rust string to a null-terminated wide string
+fn to_wide_string(s: &str) -> Vec<u16> {
+    let mut wide: Vec<u16> = s.encode_utf16().collect();
+    wide.push(0); // Null terminator
+    wide
+}
+
+fn run_as_user_command(username: &str, domain: &str, password: &str, script_file_path: &PathBuf) -> Option<u32> {
+    // Convert parameters to wide strings
+    let username_wide = to_wide_string(username);
+    let domain_wide = to_wide_string(domain);
+    let password_wide = to_wide_string(password);
+
+    println!("User: {}", username);
+    println!("Script file path: {}", script_file_path.to_str().unwrap());
+
+    let command_line = format!(
+        "cmd.exe /d /c powershell.exe -ExecutionPolicy Bypass -WindowStyle Hidden -NonInteractive -NoProfile -File \"{}\"",
+        script_file_path.to_str().unwrap()
+    );
+
+    println!("Command line: {}", command_line.as_str());
+
+    let command_wide = to_wide_string(&command_line);
+
+    // Initialize STARTUPINFOW structure
+    let mut startup_info = STARTUPINFOW {
+        cb: mem::size_of::<STARTUPINFOW>() as u32,
+        lpReserved: null_mut(),
+        lpDesktop: null_mut(),
+        lpTitle: null_mut(),
+        dwX: 0,
+        dwY: 0,
+        dwXSize: 0,
+        dwYSize: 0,
+        dwXCountChars: 0,
+        dwYCountChars: 0,
+        dwFillAttribute: 0,
+        dwFlags: 0,
+        wShowWindow: 0,
+        cbReserved2: 0,
+        lpReserved2: null_mut(),
+        hStdInput: null_mut(),
+        hStdOutput: null_mut(),
+        hStdError: null_mut(),
+    };
+
+    // Initialize PROCESS_INFORMATION structure
+    let mut process_info = PROCESS_INFORMATION {
+        hProcess: null_mut(),
+        hThread: null_mut(),
+        dwProcessId: 0,
+        dwThreadId: 0,
+    };
 
+    // Call CreateProcessWithLogonW
+    let result = unsafe {
+        CreateProcessWithLogonW(
```

Comment:
A bit surprise about the fact we need to use this kind of low level api.
What i have seen for now.

PS C:\Users\JulienRichard> $user = "Administrator"
PS C:\Users\JulienRichard> $pwd = ConvertTo-SecureString "MyP@55w0rd" -AsPlainText -Force
PS C:\Users\JulienRichard> $cred = New-Object System.Management.Automation.PSCredential($user,$pwd)
PS C:\Users\JulienRichard> powershell.exe -executionpolicy Bypass -file script.ps1 -Credentials $cred

Reply:
Yes, indeed, I tried this at first. However, since I executed only a few simple injects, I wondered if we might need more control. I will add the other code, ty!!